core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1872 @@
+    /// ```
+    pub fn into_mut(self) -> &'a mut V {
+        self.into_kv_mut()
+            .expect("every occupied entry always has a key/value pair; qed")
+            .1
+    }
+
+    /// Takes the value of the entry out of the map, and returns it.
+    ///
+    /// # Examples
+    ///
+    /// ```no_compile
+    /// use ink_core::storage::BTreeMap;
+    /// use ink_core::storage::btree_map::Entry;
+    ///
+    /// let mut map: BTreeMap<&str, usize> = new_btree_map();
+    /// map.entry("poneyland").or_insert(12);
+    ///
+    /// if let Entry::Occupied(o) = map.entry("poneyland") {
+    ///     assert_eq!(o.remove(), 12);
+    /// }
+    /// // If we try to get "poneyland"'s value, it'll panic:
+    /// // println!("{}", map["poneyland"]);
+    /// ```
+    pub fn remove(self) -> V {
+        self.tree.remove_kv(self.handle).1
+    }
+
+    /// Inserts a value into this entry.
+    fn insert(&mut self, value: V) -> Option<V> {
+        let node = self
+            .tree
+            .get_node_mut(&self.handle.into())
+            .expect("every occupied entry always belongs to a node; qed");
+        node.vals[self.handle.idx()].replace(value)
+    }
+
+    fn kv_mut(&mut self) -> Option<(&mut K, &mut V)> {
+        let idx = self.handle.idx();
+        let node = self.tree.get_node_mut(&self.handle.into())?;
+        let k = node.keys[idx].as_mut()?;
+        let v = node.vals[idx].as_mut()?;
+        Some((k, v))
+    }
+
+    fn into_kv_mut(self) -> Option<(&'a mut K, &'a mut V)> {
+        let idx = self.handle.idx();
+        let node = self.tree.get_node_mut(&self.handle.into())?;
+        let k = node.keys[idx].as_mut()?;
+        let v = node.vals[idx].as_mut()?;
+        Some((k, v))
+    }
+}
+
+/// A vacant entry of a storage map.
+pub struct VacantEntry<'a, K, V> {
+    // The `key` needs to be moved for putting, hence we have to use `Option<K`>
+    // to prevent running into partial move errors.
+    key: Option<K>,
+    tree: &'a mut BTreeMap<K, V>,
+    handle: KVHandle,
+}
+
+impl<K, V> Flush for InternalEntry<K, V>
+where
+    K: Encode + Flush,
+    V: Encode + Flush,
+{
+    #[inline]
+    fn flush(&mut self) {
+        match self {
+            InternalEntry::Vacant(vacant) => vacant.flush(),
+            InternalEntry::Occupied(occupied) => occupied.flush(),
+        }
+    }
+}
+
+/// Inserts `val` at `idx` into `slice` while shifting all subsequent items to
+/// the right by one. The last element of the slice will fall out.
+unsafe fn slice_insert<T>(slice: &mut [T], idx: usize, val: T) {
+    ptr::copy(
+        slice.as_ptr().add(idx),
+        slice.as_mut_ptr().add(idx + 1),
+        slice.len() - idx - 1,
+    );
+    ptr::write(slice.get_unchecked_mut(idx), val);
+}
+
+/// Extracts the element at `idx` from `slice` while shifting all subsequent items to
+/// the left by one.
+///
+/// Returns the extracted element.
+unsafe fn slice_remove<T>(slice: &mut [Option<T>], idx: usize) -> Option<T> {
```

Comment:
Isn't this what `Vec::remove(&mut self, index: usize)` does?
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove